### PR TITLE
Add types to TypeHash and moved away from __slots__ usage

### PR DESCRIPTION
**Pull request by InvincibleRMC**
Adds static typing to type_hash.py. In the constructor I switch away from `**kwargs` to explicit variable names. If there is some reason the `**kwargs` is preferred for now I can use Any. Just let me know so I can add a TODO to use Unpack[] when python 3.12 is supported. Or I suppose I could use typing_extensions package to use Unpack on older versions of python.

### CODE REVIEW

File: rclpy/rclpy/type_hash.py [hunk 1]
```diff
@@ -41,12 +33,12 @@ def version(self):
...
 
     @version.setter
-    def version(self, value):
+    def version(self, value: int) -> None:
         assert isinstance(value, int)
         self._version = value
 
     @property
-    def value(self):
+    def value(self) -> bytes:
```
Comment (christophebedard) [inline rclpy/rclpy/type_hash.py:41]: Should we remove the typing info from the return value documentation below?

```diff
-:rtype: bytes
```

File: rclpy/rclpy/type_hash.py [hunk 2]
```diff
@@ -18,20 +18,12 @@ class TypeHash:
...
-
-        self.version = kwargs.get('version', -1)
-        self.value = kwargs.get('value', bytes(self._TYPE_HASH_SIZE))
+    def __init__(self, version: int = -1, value: bytes = bytes(_TYPE_HASH_SIZE)):
+        self.version = version
+        self.value = value
 
     @property
-    def version(self):
+    def version(self) -> int:
```
Comment (christophebedard) [inline rclpy/rclpy/type_hash.py:26]: Should we remove the typing info from the return value documentation below?

```diff
-:rtype: int
```

**Review by fujitatomoya — APPROVED** [review]

**Review by sloretz — APPROVED** [review]

**Comment by sloretz** [comment]
CI ([repos file]() build: `--packages-up-to rclpy` test: ` --packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20448/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14850/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21191/)

**Comment by clalancette** [comment]
> CI ([repos file](https://gist.githubusercontent.com/sloretz/fd31d2745812ae359c759244258b243c/raw/8b6cee727c30fdaa7f48c631fed5c65979dc06bf/ros2.repos) build: `--packages-up-to rclpy` test: ` --packages-select rclpy`)

It turns out that this isn't enough CI; this is causing failures in CI now: https://ci.ros2.org/job/ci_linux/20452/testReport/junit/ros2cli.ros2cli.test/test_daemon/test_get_subscriptions_info_by_topic/ .

I'm going to propose a revert here.

**Pull request merged**